File: Bwl.Imaging/Geometry/Parameters/ParametersDictionary.vb
Public Class ParametersDictionary
    Implements IEnumerable(Of Parameter)

    Private _data As New Dictionary(Of String, Parameter)
    Private _syncRoot As New Object

    Public Sub New()
    End Sub

    Public Sub New(values As IEnumerable(Of Parameter))
        Me.AddRange(values)
    End Sub

    Public Overridable Function GetEnumerator() As IEnumerator(Of Parameter) Implements IEnumerable(Of Parameter).GetEnumerator
        Dim values As IEnumerable(Of Parameter) = Nothing
        SyncLock _syncRoot
            values = _data.Values.ToArray()
        End SyncLock
        Return values.GetEnumerator()
    End Function

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Throw New NotImplementedException()
    End Function

    Default Public Overloads Property Item(key As String) As Parameter
        Get
            SyncLock _syncRoot
                Return _data(key)
            End SyncLock
        End Get
        Set(value As Parameter)
            SyncLock _syncRoot
                _data(key) = value
            End SyncLock
        End Set
    End Property

    Public Function ItemValue(key As String) As String
        SyncLock _syncRoot
            Return _data(key).Value
        End SyncLock
    End Function

    Public Overloads Sub Add(key As String, value As String, Optional visible As Boolean = False)
        SyncLock _syncRoot
            _data.Add(key, New Parameter(key, value, visible))
        End SyncLock
    End Sub

    Public Overloads Sub Add(value As Parameter)
        SyncLock _syncRoot
            _data.Add(value.Key, value)
        End SyncLock
    End Sub

    Public Sub AddRange(values As IEnumerable(Of Parameter))
        SyncLock _syncRoot
            For Each value In values
                _data.Add(value.Key, value)
            Next
        End SyncLock
    End Sub

    Public Sub Remove(key As String)
        SyncLock _syncRoot
            _data.Remove(key)
        End SyncLock
    End Sub

    Public Function ContainsKey(key As String) As Boolean
        SyncLock _syncRoot
            Return _data.ContainsKey(key)
        End SyncLock
    End Function

    Public Overloads Sub Clear()
        SyncLock _syncRoot
            _data.Clear()
        End SyncLock
    End Sub

    Public Function ToArray() As Parameter()
        SyncLock _syncRoot
            Return _data.Values.ToArray()
        End SyncLock
    End Function
End Class

File: Bwl.Imaging.Skia/Geometry/Parameters/ParametersDictionary.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Bwl.Imaging
{

    [JsonConverter(typeof(ParametersDictionaryJsonConverter))]
    public class ParametersDictionary : IEnumerable<Parameter>
    {

        private Dictionary<string, Parameter> _data = new Dictionary<string, Parameter>();
        private object _syncRoot = new object();

        public ParametersDictionary()
        {
        }

        public ParametersDictionary(IEnumerable<Parameter> values)
        {
            AddRange(values);
        }

        public virtual IEnumerator<Parameter> GetEnumerator()
        {
            IEnumerable<Parameter> values = null;
            lock (_syncRoot)
                values = _data.Values.ToArray();
            return values.GetEnumerator();
        }

        private IEnumerator IEnumerable_GetEnumerator()
        {
            throw new NotImplementedException();
        }

        IEnumerator IEnumerable.GetEnumerator() => IEnumerable_GetEnumerator();

        public Parameter this[string key]
        {
            get
            {
                lock (_syncRoot)
                    return _data[key];
            }
            set
            {
                lock (_syncRoot)
                    _data[key] = value;
            }
        }

        public string ItemValue(string key)
        {
            lock (_syncRoot)
                return _data[key].Value;
        }

        public void Add(string key, string value, bool visible = false)
        {
            lock (_syncRoot)
                _data.Add(key, new Parameter(key, value, visible));
        }

        public void Add(Parameter value)
        {
            lock (_syncRoot)
                _data.Add(value.Key, value);
        }

        public void AddRange(IEnumerable<Parameter> values)
        {
            lock (_syncRoot)
            {
                foreach (var value in values)
                    _data.Add(value.Key, value);
            }
        }

        public void Remove(string key)
        {
            lock (_syncRoot)
                _data.Remove(key);
        }

        public bool ContainsKey(string key)
        {
            lock (_syncRoot)
                return _data.ContainsKey(key);
        }

        public void Clear()
        {
            lock (_syncRoot)
                _data.Clear();
        }

        public Parameter[] ToArray()
        {
            lock (_syncRoot)
                return _data.Values.ToArray();
        }
    }

    internal sealed class ParametersDictionaryJsonConverter : JsonConverter<ParametersDictionary>
    {
        public override ParametersDictionary Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            var map = JsonSerializer.Deserialize<Dictionary<string, Parameter>>(ref reader, options)
                      ?? new Dictionary<string, Parameter>();

            var result = new ParametersDictionary();
            foreach (var kv in map)
                result.Add(kv.Value ?? new Parameter(kv.Key, string.Empty));
            return result;
        }

        public override void Write(Utf8JsonWriter writer, ParametersDictionary value, JsonSerializerOptions options)
        {
            var map = value.ToArray().ToDictionary(p => p.Key, p => p);
            JsonSerializer.Serialize(writer, map, options);
        }
    }
}